Modules/Common/CardanoLikeNativeTokensModule.php

```diff
@@ -0,0 +1,321 @@
+                    FROM tx_out
+                    JOIN ma_tx_out ON (ma_tx_out.tx_out_id = tx_out.id)
+                    JOIN multi_asset ON ma_tx_out.ident = multi_asset.id
+                    WHERE tx_out.tx_id in ({$in_query})
+                    ORDER BY tx_out.tx_id, tx_out.index"));
+
+        foreach ($outs as $out)
+        {
+            $block[($out['tx_id'])]['outputs'][] = $out;
+        }
+
+        // inputs - due to the way foreign keys are chosen in relevant tables, 3-table-join ends up 8-10 times slower then this:
+
+        $ins_1 = pg_fetch_all(pg_query($this->db,
+            "SELECT tx_in.tx_in_id, tx_out.address, tx_in.id as orderby, tx_out.id as nextkey
+                    FROM tx_in
+                    JOIN tx_out ON (tx_in.tx_out_id = tx_out.tx_id AND tx_in.tx_out_index = tx_out.index)
+                    WHERE tx_in.tx_in_id IN ({$in_query})"));
+
+        if (count($ins_1) < 1)
+        {
+            $this->set_return_events([]);
+            $this->set_return_currencies([]);
+            return;
+        }
+
+        $ins_dict = [];
+        $subquery = [];
+
+        foreach ($ins_1 as $aux)
+        {
+                $ins_dict[$aux['nextkey']] = $aux;
+                $subquery[] = $aux['nextkey'];
+        }
+        $subquery = implode(', ', $subquery);
+
+        $ins_2 = pg_fetch_all(pg_query($this->db,
+            "SELECT ma_tx_out.tx_out_id as id, multi_asset.fingerprint, ident AS token, ma_tx_out.quantity
+                    FROM ma_tx_out
+                    JOIN multi_asset ON ma_tx_out.ident = multi_asset.id
+                    WHERE tx_out_id in ({$subquery})"));
+
+        $ins = [];
+        foreach ($ins_2 as $aux)
+        {
+            $ins[] = [
+                'tx_in_id'    => $ins_dict[$aux['id']]['tx_in_id'],
+                'address'     => $ins_dict[$aux['id']]['address'],
+                'orderby'     => $ins_dict[$aux['id']]['orderby'],
+                'fingerprint' => $aux['fingerprint'],
+                'token'       => $aux['token'],
+                'quantity'    => $aux['quantity'],
+            ];
+        }
+        usort($ins, fn($a, $b) => $a['orderby'] <=> $b['orderby']);
+
+        unset($ins_1);
+        unset($ins_2);
+        unset($ins_dict);
+        unset($subquery);
+
+        // this works under assumption that one cannot transfer tokens without transfering ada (ie MA-outs are strictly a subset of ADA-outs)
+
+        foreach ($ins as $in)
+        {
+            $block[($in['tx_in_id'])]['inputs'][] = $in;
+        }
+
+        $deltas = pg_fetch_all(pg_query($this->db,
+            "SELECT ma_tx_mint.tx_id, multi_asset.fingerprint, ident AS token, ma_tx_mint.quantity
+                    FROM ma_tx_mint
+                    JOIN multi_asset ON ma_tx_mint.ident = multi_asset.id
+                    WHERE tx_id IN ({$in_query})
+                    ORDER BY ma_tx_mint.id"));
```

Comment:
Is this correct ordering?

Reply:
As far as I understand, there cannot be more than 1 mint/burn per transaction so it is. Though, it shouldn't hurt to enforce extra safety. Will do.